## TCRsep: T-cell receptor selection estimation procedure
TCRsep is a python software for the inference of the selection factor for immune receptor repertoires. It takes a productive TCR repertoire and pre-selection repertoire (optional) as inputs for model training. After that, it outputs the selection factors for any given TCR clonetypes (defined in the __CDR3-V-J__ format). It also outputs their post-selection probabilities that can be further utilized to analyze the receptor sharing pattern and identify indicative disease-associated receptors. 
 <br />

<img src="https://github.com/jiangdada1221/TCRsep/blob/main/figs/workflow_github.png" width="800"> <br />

## Installation
TCRsep is available on PyPI and can be installed via pip: <br />
 ```pip install tcrsep``` <br /> <br />
Or install TCRsep via Github: <br />
```
git clone https://github.com/jiangdada1221/TCRsep.git
cd TCRsep
pip install .
```
TCRsep depends on multiple packages. It's highly recommended to manually install `torch` first. If the installation fails, make sure that the following dependencies are installed correctly: <br /> 
`torch` >= 1.8.0 (Tested on torch [1.8.0+cuda11.1](https://pytorch.org/get-started/previous-versions/#v180))<br />
[`olga`](https://github.com/statbiophys/OLGA) (For modeling the generation of TCR) <br />
[`tcr2vec`](https://github.com/jiangdada1221/TCR2vec) (For embedding TCR) <br /> 

## Usage instructions
#### 1. Train a TCRsep model: 
type `python train.py -h` to display all the commandline options: 
|Commands|Description|
|--|--|
|`-h, --help`|show the help message and exit|
|`--post_data_path=FILE`|(__required__) The path to the input repertoire file (or its embedding file). An example: `data/example.csv` (`data/example_emb.npy.gz`). If given the embedding path, will save the time to embed input TCRs.| 
|`--save_dir=DIR`|The directory that saves all the output files. Default `result/`.|  
|`--pre_data_path=FILE`|The path to the pre-selection repertoire file (or embedding). If not specified, pre-selection TCRs will be automatically generated by the generation model incorporated in TCRsep.|
|`--gen_model_path=DIR`|The path to the generation model. If not specified, will use the default generation model inferred on Emerson data.|
|`--iters=NUM`|Iterations for the training process. Default 10,000.|
|`--alpha=NUM`|The parameter Œ±. Recommended using the default value 0.1.|
|`--dropout=NUM`|The dropout rate. Default 0.1.|                       
|`--batchsize=NUM`|Batch size. Default 1,000.|
|`--val_ratio=NUM`|Fraction of the data serving as validation. Default 0.1.|  
|`--simulation`|Set to True in simulation experiments. Default False.|

__Notes:__ the data file (`.csv/.tsv`) needs at least three columns specifying the CDR3Œ≤ amino acid sequences, V genes, and J genes: `CDR3.beta`, `V` and `J`. The `save_dir` will contain the pre-selection repertoire file, embeddings of pre- and post-selection repertoires, the selection model, and a json file recording the input arguments.  

#### 2. Use the TCRsep to infer selection factors, pre- and post-selection probabilities:
type `python eval.py -h` to display all the commandline options:
|Commands|Description|
|--|--|
|`--data_path=FILE`|(__required__) The path to the query repertoire file that needs to be evaluated. An example: `data/query_data.csv`.| 
|`--data_emb_path=FILE`|The path to the embedding of the query repertoire. If specified, will save the time for embedding.|  
|`--sel_model_path=FILE`|The path to the TCRsep model. If not specified, will use the default model inferred on Emerson data.|  
|`--save_dir=DIR`|The directory that saves all the output files. Default `result_eval/`.|  
|`--gen_model_path=DIR`|The path to the generation model. If not specified, will use the default generation model inferred on Emerson data.|
|`--alpha=NUM`|The parameter Œ± of TCRsep. Default 0.1.|                       
|`--simulation`|Set to True in simulation experiments. Default False.|

#### 3 Usages of TCRsep in Python script
3.1 Utilities of TCRsep module:
```python
from tcrsep.estimator import TCRsep
sel_model = TCRsep(default_sel_model=True)
query_tcrs = [['CASSLGAGGSGTEAFF','TRBV7-9','TRBJ1-1'], ['CASTKAGGSSYEQYF','TRBV6-5','TRBJ2-7']]
sel_factors = sel_model.predict_weights(query_tcrs) #obtain selection factors
pgens, pposts = sel_model.get_prob(query_tcrs) #obtain pre- and post-selection probs 

# draw samples from p_post
post_samples = sel_model.sample(n=10)
```
3.2 Sharing analysis by TCRsep:
```python
from tcrsep.sharing_analysis import Sharing, DATCR
sharing_predictor = Sharing('data/sharing')

# predict sharing numbers of TCRs in query_data.csv among reps in the folder, "data/sharing"
sharing_pre,sharing_real = sharing_predictor.predict_sharing('data/query_data_evaled.csv') 

# predict the sharing spectrum for reps in "data/sharing"
spectrum_pre,spectrum_real = sharing_predictor.sharing_spectrum(est_num=100000) 

# identify DATCRs
DATCR_predictor = DATCR('data/sharing')
pvalues = DATCR_predictor.pvalue('data/query_data_evaled.csv')
```
__Notes:__ the `query_data_evaled.csv` should contain an additional column `ppost` specifying the post-selection probabilities. Please [eval](https://github.com/jiangdada1221/TCRsep?tab=readme-ov-file#2-use-the-tcrsep-to-infer-selection-factors-pre--and-post-selection-probabilities) the `query_data` file first if it only contains the `CDR3-V-J` information.  

#### 4 Simulate synthetic datasets with known selection factors
type `python rejection_sampler.py -h` to display all the commandline options: 
|Commands|Description|
|--|--|
|`-h, --help`|show the help message and exit|
|`--simu_model=STRING`| Type of simulation experiments. Possible inputs: `V_select`/`J_select`/`VJ_select`/`Motif_select`|
|`--gen_model_path=DIR`|The path to the generation model. If not specified, will use the default generation model.|   
|`--pre_data_path=FILE`|The path to the pre-sel repertoire file. If not specified, pre-sel TCRs will be automatically generated by the generation model. The pre-sel TCRs are used to compute statistics of the pre-sel repertoire.|
|`--save_path=FILE`|The path to the file that saves the simulated post-sel repertoire. Default `data/simu.csv`.|
|`--sample_num=NUM`|The number of simulated TCRs.|
|`--temp=NUM`|The temperature parameter controling the sharpness of simulated selection factors in gene-related simulations. Default value 0.1.|
|`--seed=NUM`|The seed controling the randomness. Default 43.|
|Below are args specific for `Motif_select`|The following args will not impact Gene-related simulations. |                       
|`--start_pos=NUM`|The starting position of candiate k-mers within CDR3 region. Default 3.|
|`--k=NUM`|The length of selected motif. Default 3.|  
|`--sel_pos=NUM`|Select the k-mer with certain frequency rank as the chosen motif. Possible values: `0`:the most frequent one; `1`:the medium one; `2`: the one with lowest frequency. Default 0.|
|`--fre_thre=NUM`|The frequency threshold for candiate k-mers. Default 0.005.|
|`--sel_factor_rel=NUM`|Relative selection factor of the chosen k-mer over other k-mers. Default 2.|

For other customized usages, we provide the following module descriptions. Users can refer to the docstrings included in the script files for more details.

| Script name                                    | Usage                                              |    
|------------------------------------------------|----------------------------------------------------|
| estimator.py                              | The TCRsep module                                  |
| train.py                                  | Train TCRsep by command line                       |
| eval.py                                   | Inference of sel_factors and probs by command line |
| pgen.py                                   | The generation module that relies on the OLGA package|
| rejection_sampler.py                      | Sampling module that enable to draw samples from P_post      |
| sharing_analysis.py                       | Modules used for sharing analysis                      |
| utils.py                                  | N/A (contains utility functions)                  |

## Processed data
All the data used in the manuscript is publicly available, so we suggest readers refer to the original papers for more details. We also provide our processed data which can be publicly [downloaded](https://drive.google.com/file/d/1C_owgLXH9Ywx8B1_0KQASZNqu26Kd0aT/view?usp=sharing).

__Data details__:
- üìÅ `data/` 
  - üìÅ `Chu_reps/`
  - üìÅ `CMV_reps/`
    - üìÅ `HIP_batch/`
    - üìÅ `Keck_batch/`
  - üìÅ `COVID19_reps/`
    - üìÅ `Capelle_data/`
    - üìÅ `ImmuneCODE/`
    - üìÅ `ImmuneCODE_release2/`
  - üìÅ `nonbinding_TCRs/`
    - `10x_nonbinding_TCRs` Nonbinding TCRs extracted from 10x genomics. 
  - üìÅ `specific_TCRs/`
    - `CMV_asso_emerson.csv` CMV associated TCRs identified in Emerson et al. 2017.
    - `mira.csv` COVID-19 specific TCRs identied by MIRA expeirment in the ImmuneCODE project.
    - `vdjdb_cmv.csv` CMV specific TCRs extracted from VDJdb
    - `vdjdb_covid19.csv` YLQPRTFLL specific TCRs extracted from VDJdb
  - üìÅ `simulation_reps/`  

Note that the repertoire files are included in the `reps/` directories (e.g. `Chu_reps/`).

## Contact
```
Author: Yuepeng Jiang
Email: yuepjiang3-c@my.cityu.edu.hk/yuj009@eng.ucsd.edu/jiangdada12344321@gmail.com
Note: For instant query, feel free to send me an email since I check email often. 
Otherwise, you may open an issue section in this repository.
```

## License

Free use of TCRsep is granted under the terms of the MIT License.

<!-- ## Citation 
```

``` -->
